flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java

```diff
@@ -104,22 +105,15 @@ public void applyProjection(int[][] projectFields) {
         .tableLoader(loader)
         .properties(properties)
         .project(getProjectedSchema())
+        .projectedFields(projectedFields)
         .limit(limit)
         .filters(filters)
         .flinkConf(readableConfig)
         .build();
   }
 
   private TableSchema getProjectedSchema() {
-    if (projectedFields == null) {
-      return schema;
-    } else {
-      String[] fullNames = schema.getFieldNames();
-      DataType[] fullTypes = schema.getFieldDataTypes();
-      return TableSchema.builder().fields(
-          Arrays.stream(projectedFields).mapToObj(i -> fullNames[i]).toArray(String[]::new),
-          Arrays.stream(projectedFields).mapToObj(i -> fullTypes[i]).toArray(DataType[]::new)).build();
-    }
+    return projectedFields == null ? schema : projectSchema(schema, topProjectedFields);
```

Comment:
It was revised based on the comments above. In the new version, it pushes the projection down.  top-level projected is no longer used so that additional fields are not retrieved